### ISSUE TITLE
Using `as` to get `todo!` to work with `impl` return gives false unreachable code warning

### ISSUE THREAD

**Issue by ChaiTRex** [rust-lang/rust#123871]
### Code

```Rust
pub fn f() -> impl Iterator<Item = ()> {
    todo!() as core::iter::Empty<()>
}
```

### Current output

```Shell
warning: unreachable expression
 --> src/lib.rs:2:5
  |
2 |     todo!() as core::iter::Empty<()>
  |     -------^^^^^^^^^^^^^^^^^^^^^^^^^
  |     |
  |     unreachable expression
  |     any code following this expression is unreachable
  |
  = note: `#[warn(unreachable_code)]` on by default

warning: `playground` (lib) generated 1 warning
```

### Desired output

```Shell
No warning
```

### Rationale and extra context

Using `todo!()` with a return type that is `impl` currently causes a type error. Using `as` to give it a specific type fixes that but brings this incorrect warning message.

`as T` by itself shouldn't really be considered an expression.

### Rust Version

```Shell
rustc 1.77.2 (25ef9e3d8 2024-04-09)
binary: rustc
commit-hash: 25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04
commit-date: 2024-04-09
host: aarch64-apple-darwin
release: 1.77.2
LLVM version: 17.0.6

rustc 1.78.0-beta.6 (27011d5dc 2024-04-09)
binary: rustc
commit-hash: 27011d5dc09a8d1ac2289d024f54ecac9fc70d47
commit-date: 2024-04-09
host: aarch64-apple-darwin
release: 1.78.0-beta.6
LLVM version: 18.1.2

rustc 1.79.0-nightly (aa067fb98 2024-04-10)
binary: rustc
commit-hash: aa067fb984d36462548bb785da221bfaf38253f0
commit-date: 2024-04-10
host: aarch64-apple-darwin
release: 1.79.0-nightly
LLVM version: 18.1.3
```

**Comment by cuviper** [comment]
> `as T` by itself shouldn't really be considered an expression.

It's literally a [type cast expression](https://doc.rust-lang.org/reference/expressions/operator-expr.html#type-cast-expressions).

If we had [type ascription](https://github.com/rust-lang/rfcs/pull/2522), then I would agree *that* shouldn't look unreachable.

**Comment by ChaiTRex** [comment]
OK, then perhaps the problem is that `todo!() as T` should not be considered unreachable as if it comes after a statement that panics or returns or whatever, as it doesn't come after such an expression. It is that expression.

**Comment by cuviper** [comment]
I would say that the panicking `todo!()` is a *sub*expression of the `todo!() as T` expression, and it does render the rest of the expression unreachable, just like if you had `todo!() + 1` or `foo(todo!())`.

Don't get me wrong though, I am sympathetic to the need for a type in `todo!()` with `-> impl Trait`. We just don't have a good way to write that yet.

**Comment by cuviper** [comment]
For iterators, if you're willing to have a lazy panic until the iterator is actually used, then this isn't too bad:

```rust
pub fn f() -> impl Iterator<Item = ()> {
    core::iter::repeat_with(|| todo!())
}
```

**Comment by eggyal** [comment]
[Relevant recent discussion on IRLO](https://internals.rust-lang.org/t/todo-should-prevent-unused-variables-warning/20542/24).

**Comment by ryanavella** [comment]
> For iterators, if you're willing to have a lazy panic until the iterator is actually used, then this isn't too bad:

I do this for an eager panic:

```rust
pub fn f() -> impl Iterator<Item = ()> {
    todo!();
    None.into_iter()
}
```

The unreachable warning might bother some people, but I prefer the warning as a reminder that it is unfinished.
